### ISSUE TITLE
migrating from 2.0.12 to latest (3.1.2) : linking produces duplicate symbols

### ISSUE THREAD

**Issue by flatmax** [emscripten-core/emsdk#1194]
I'm trying to update my system from 2.0.12 to latest (3.1.2). I've found I have to switch installation prefixes to sysroot (https://github.com/emscripten-core/emsdk/issues/1193).

Now I'm finding strangitude in the linking of many libraries together. For some reason duplicate symbols are found :
```
binding libdsp non WASM
emcc: warning: --llvm-opts is deprecated.  All non-emcc args are passed through to clang. [-Wdeprecated]
wasm-ld: error: duplicate symbol: memmove
>>> defined in .libs/libdsp.so
>>> defined in /usr/local/emsdk/upstream/emscripten/cache/sysroot/lib/libFLAC.so

wasm-ld: error: duplicate symbol: memset
>>> defined in .libs/libdsp.so
>>> defined in /usr/local/emsdk/upstream/emscripten/cache/sysroot/lib/libFLAC.so

make[1]: *** [Makefile:1127: all-local] Error 1
make[1]: Leaving directory 'gtkiostream/src'

```

**Comment by sbc100** [comment]
Its hard to say what is going on there without seeing how those two libraries are getting built.

Can you figure out why `memmove` is being defined in those two libraries?    I think having `memmove` defined in both of them should have been and error both in 2.0.12 and in 3.1.2.

Can you track down where the `memmove` definition is coming from?   Perhaps you could share the full link command used to generate  libFLAC.so?   What happens if you add `-Wl,--trace-symbol=memmove` to this command?

**Comment by sbc100** [comment]
BTW, these `.so` files are not real shared shared libraries under emscripten, and I strongly recommend swithing from fake shared library support to just using static libraries.  This will improve the error messages you get here, and might actually solve the issue.   

Since it looks like you are using autoconf I think the simplest way to force static libraries is to pass `--disable-shared` to configure (at least this works for some projects I think).

**Comment by flatmax** [comment]
ok - making compilations native imrpved things somewhat. A new problem appears with importing the module like so :

import libTest from './libTest.js';

Error while transforming libTest.js: Could not resolve import "module".

Any advice ? This is the bind command :

	@emcc --bind --llvm-opts 1 --memory-init-file 0 -s SINGLE_FILE=1 -s EXPORT_ES6=1 -s "MODULARIZE=1" -s EXPORT_NAME="'libTest'" -s "BINARYEN_METHOD='native-wasm'" -s "WASM=1" -s ALLOW_MEMORY_GROWTH=1 $(AM_CXXFLAGS) .libs/libTestWASM.so -o .libs/libTest.js

**Comment by sbc100** [comment]
I'm not familiar with that error message.  What tools is producing it?  Does it not give you a line number of anything like that?

**Comment by flatmax** [comment]
Nope - strange eh !

Could you give any suggestions to clean up my emcc bind line. Also is my import correct ?

I've tried this :
import { libTest } from './libTest.js';

```
@emcc -lembind -O2 --memory-init-file 0 -sEXPORT_ES6=1 -s MODULARIZE=1 -s EXPORT_NAME="'libTest'" -s SINGLE_FILE=1 -s "BINARYEN_METHOD='native-wasm'" -s WASM=1 -s ALLOW_MEMORY_GROWTH=1 $(AM_CXXFLAGS) .libs/libTest.so $(GTKIOSTREAM_LIBS) -o .libs/libTest.js
```

**Comment by sbc100** [comment]
Perhaps try without `SINGLE_FILE`, I'm not sure we test that with `EXPORT_EX6`?

You won't need `--memory-init-file 0`, `-s WASM=1` or `BINARYEN_METHOD` anymore.  Those are all the defaults.

Also you don't need quotes around any those options, or `=1` after them, or a space between the `-s` and the name of the option.  i.e.  just `-sMODULARIZE` rather than `-s MODULARIZE=1`.

**Comment by flatmax** [comment]
ok - I tried that - no success so far.

here are my  AM_CXXFLAGS , can you see anything there ?
```
AC_SUBST([AM_CXXFLAGS], ["-s ERROR_ON_UNDEFINED_SYMBOLS=0 -s DEMANGLE_SUPPORT=1 -std=c++11 -O3 -fno-exceptions -s NO_EXIT_RUNTIME=1 -U EIGEN_DEBUG_ASSIGN"])
```

**Comment by sbc100** [comment]
What error are you seeing now?  

Presumably not an es6 import error since you are no longer building with EXPORT_ES6? 

I would advise against `ERROR_ON_UNDEFINED_SYMBOLS` that will likely lead to missing symbols at runtime.. also `NO_EXIT_RUNTIME` is the default so you don't need that.

**Comment by flatmax** [comment]
Yes that is setup too. To assist with this debug, I have updated a generic C++ WASM repo : https://github.com/flatmax/WASM

This repo experiences the same problems with the browser loading, however the nodejs example works fine.

The C++ compile flags are set here : https://github.com/flatmax/WASM/blob/master/configure.ac#L65
```
AC_SUBST([AM_CXXFLAGS], ["-s ERROR_ON_UNDEFINED_SYMBOLS=0 -s DEMANGLE_SUPPORT=1 -std=c++11 -O3 -fno-exceptions -s NO_EXIT_RUNTIME=1 "])
```

The emscripten binding is done here : https://github.com/flatmax/WASM/blob/master/src/Makefile.am#L36-L43

For nodejs - which works fine : 
```
@emcc --bind -sALLOW_MEMORY_GROWTH=1 -sMODULARIZE -sEXPORT_NAME="'wasm'" -sWASM=0 --memory-init-file 0 $(AM_CXXFLAGS) .libs/libwasm.so -o .libs/libwasm.js
```

For the browser (which doesn't work) :
```
@emcc --bind -sALLOW_MEMORY_GROWTH=1 -sMODULARIZE -sEXPORT_NAME="'libwasm'" -sEXPORT_ES6=1 -sSINGLE_FILE=1 $(AM_CXXFLAGS) .libs/libwasm.so -o .libs/libwasm.js
```

The error is still :
```
Error while transforming libwasm.js: Could not resolve import "module".
```
This is the import statement : https://github.com/flatmax/WASM/blob/master/webcomponent/libWASM.js#L2
```
import { libwasm } from './libwasm.js';
```

**Comment by flatmax** [comment]
Ok I have tested the 3.1.* releases and this module loading problem appears in release 3.1.27.
In release 3.1.26 the module can be loaded in the webcomponent like so : https://github.com/flatmax/WASM/blob/master/webcomponent/libWASM.js#L2
```
let modProm = import('./libwasm.js'); 
```

Unfortunately I can't find the Test C++ class in the wasm code. Not sure why yet.

**Comment by flatmax** [comment]
to do your own tests, it is like so :
```
# install and actiavate an emscripten release
./embuild.sh # this will clean, configure and build the wasm library
cd webcomponent
npm i
npm start
```

**Comment by flatmax** [comment]
Now that I can load and run an ES6 module in the browser (https://github.com/emscripten-core/emsdk/issues/1198) I can confirm that the following emscripten version breaks ES6 module imports :
```
v3.1.27
```

**Comment by sbc100** [comment]
Closing in favor of #1198)
